ui/ducks/ens.js

```diff
@@ -0,0 +1,187 @@
+import { createSlice } from '@reduxjs/toolkit';
+import ENS from 'ethjs-ens';
+import log from 'loglevel';
+import networkMap from 'ethereum-ens-network-map';
+import { isConfusing } from 'unicode-confusables';
+import { isHexString } from 'ethereumjs-util';
+
+import { getCurrentChainId } from '../selectors';
+import {
+  CHAIN_ID_TO_NETWORK_ID_MAP,
+  MAINNET_NETWORK_ID,
+} from '../../shared/constants/network';
+import {
+  CONFUSING_ENS_ERROR,
+  ENS_ILLEGAL_CHARACTER,
+  ENS_NOT_FOUND_ON_NETWORK,
+  ENS_NOT_SUPPORTED_ON_NETWORK,
+  ENS_NO_ADDRESS_FOR_NAME,
+  ENS_REGISTRATION_ERROR,
+  ENS_UNKNOWN_ERROR,
+} from '../pages/send/send.constants';
+import { isValidDomainName } from '../helpers/utils/util';
+import { CHAIN_CHANGED } from '../store/actionConstants';
+import {
+  BURN_ADDRESS,
+  isBurnAddress,
+  isValidHexAddress,
+} from '../../shared/modules/hexstring-utils';
+
+// Local Constants
+const ZERO_X_ERROR_ADDRESS = '0x';
+
+const initialState = {
+  stage: 'UNINITIALIZED',
+  resolution: null,
+  error: null,
+  warning: null,
+  network: null,
+};
+
+export const ensInitialState = initialState;
+
+const name = 'ENS';
+
+let ens = null;
+
+const slice = createSlice({
+  name,
+  initialState,
+  reducers: {
+    processEnsError: (state, action) => {
+      const { ensName, network, reason } = action.payload;
+      if (
+        isValidDomainName(ensName) &&
+        reason.message === 'ENS name not defined.'
+      ) {
+        state.error =
+          network === MAINNET_NETWORK_ID
+            ? ENS_NO_ADDRESS_FOR_NAME
+            : ENS_NOT_FOUND_ON_NETWORK;
+      } else if (reason.message === 'Illegal Character for ENS.') {
+        state.error = ENS_ILLEGAL_CHARACTER;
+      } else {
+        log.error(reason);
+        state.error = ENS_UNKNOWN_ERROR;
+      }
+    },
+    processEnsResult: (state, action) => {
```

Comment:
I noticed that there are many branches where the `error`, `resolution`, and `warning` state are not set. They would retain their old value, even though it was set for a previous lookup.

With both this function and `processEnsResult`, we want to update state based upon the most recent ENS lookup and discard state from the previous lookup. It might make sense to combine them into one reducer function (especially since this one is already setting the error state anyway), and ensure the ENS lookup-related state is overwritten on every branch.

Reply:
Maybe the name `ensLookup` would be better as well? :thinking: In the spirit of naming these after the events they're meant to correspond to, rather than as verbs describing what they do.

Reply:
I'm going to merge in the qa fixes branch soon which resolves some of this, then we can explore changes.

Reply:
@Gudahtt done, could you take another look and then we can discuss improvements?

Reply:
I think my comment still applies? The old state is replaced in a few more branches now, but not all of them.

Reply:
Done in https://github.com/MetaMask/metamask-extension/pull/10965/commits/92fd02bb31c9a87b4ef78c5ab8cf27fca2803b71